Mathlib/CategoryTheory/Sites/Equivalence.lean

```diff
@@ -0,0 +1,214 @@
+  NatIso.ofComponents (fun F ↦ ⟨⟨(isoWhiskerRight e.op.counitIso F.val).hom⟩,
+    ⟨(isoWhiskerRight e.op.counitIso F.val).inv⟩,
+    Sheaf.hom_ext _ _ (isoWhiskerRight e.op.counitIso F.val).hom_inv_id,
+    Sheaf.hom_ext _ _ (isoWhiskerRight e.op.counitIso F.val).inv_hom_id⟩ ) (by aesop)
+
+/-- The equivalence of sheaf categories. -/
+def sheafCongr : Sheaf J A ≌ Sheaf (e.locallyCoverDense J).inducedTopology A where
+  functor := e.sheafCongr_functor J
+  inverse := e.sheafCongr_inverse J
+  unitIso := e.sheafCongr_unitIso J
+  counitIso := e.sheafCongr_counitIso J
+  functor_unitIso_comp X := by
+    ext
+    simp only [id_obj, sheafCongr_functor_obj_val_obj, comp_obj, Sheaf.instCategorySheaf_comp_val,
+      NatTrans.comp_app, sheafCongr_inverse_obj_val_obj, Opposite.unop_op,
+      sheafCongr_functor_map_val_app, sheafCongr_unitIso_hom_app_val_app,
+      sheafCongr_counitIso_hom_app_val_app, sheafCongr_functor_obj_val_map, Quiver.Hom.unop_op,
+      Sheaf.instCategorySheaf_id_val, NatTrans.id_app]
+    simp [← Functor.map_comp, ← op_comp]
+
+/-- This would allow to weaken the assumption `HasLimits A`. -/
+proof_wanted hasMultiEqualizer_index_of_equiv
+    [∀ (P : Cᵒᵖ ⥤ A) (X : C) (S : J.Cover X), HasMultiequalizer (S.index P)]
+    (P : Dᵒᵖ ⥤ A) (X : D) (S : (e.locallyCoverDense J).inducedTopology.Cover X) :
+    HasMultiequalizer (S.index P)
+
+/-- This would allow to weaken the assumption `HasColimits A`. -/
+proof_wanted hasColimitsOfShape_cover_of_equiv
+    [∀ (X : C), HasColimitsOfShape (J.Cover X)ᵒᵖ A] (X : D) :
+    HasColimitsOfShape ((e.locallyCoverDense J).inducedTopology.Cover X)ᵒᵖ A
+
+variable [HasLimits A] [HasColimits A]
+
+variable [ConcreteCategory A]
+
+/-- This would allow to weaken the assumption `PreservesFilteredColimits (forget A)`. -/
+proof_wanted preservesColimitsOfShape_cover
+    [∀ (X : C), PreservesColimitsOfShape (J.Cover X)ᵒᵖ (forget A)] (X : D) :
+    Nonempty (PreservesColimitsOfShape
+      ((e.locallyCoverDense J).inducedTopology.Cover X)ᵒᵖ (forget A))
+
+variable [PreservesLimits (forget A)] [ReflectsIsomorphisms (forget A)]
+  [PreservesFilteredColimits (forget A)]
+
+/-- Transport a preasheaf to the equivalent category and sheafify there. -/
+noncomputable
+def transportAndSheafify : (Cᵒᵖ ⥤ A) ⥤ Sheaf J A :=
+  e.op.congrLeft.functor ⋙ presheafToSheaf _ _ ⋙ (e.sheafCongr J).inverse
+
+/-- An auxiliary definition for the sheafification adjunction. -/
+noncomputable
+def transportIsoSheafToPresheaf : (e.sheafCongr J).functor ⋙
+    sheafToPresheaf (e.locallyCoverDense J).inducedTopology A ⋙
+    e.op.congrLeft.inverse ≅ sheafToPresheaf J A :=
+  NatIso.ofComponents (fun F ↦ isoWhiskerRight e.op.unitIso.symm F.val)
+    (by intros; ext; simp [Equivalence.sheafCongr])
+
+/-- Transporting and sheafifying is left adjoint to taking the underlying presheaf. -/
+noncomputable
+def transportSheafificationAdjunction : transportAndSheafify J e ⊣ sheafToPresheaf J A :=
```

Comment:
These results are great, but is not it possible to consider introducing a type class saying that `sheafToPresheaf J A` has a left adjoint? and then redefine `presheafToSheaf` as a choice of left adjoint?

Reply:
That's indeed a better philosophy. I'll try it out!